File: nova/mobile/automation/action_executor.py
"""ActionExecutor — Runs automation actions sequentially via CommandEngine."""

import time
import logging
from typing import List, Optional
from nova.mobile.automation.models import AutomationAction
from nova.mobile.automation.enums import ActionType

logger = logging.getLogger("nova.mobile.automation.action_executor")


ACTION_TEXT_MAP = {
    ActionType.FLASHLIGHT_ON:     lambda p: "Turn on flashlight",
    ActionType.FLASHLIGHT_OFF:    lambda p: "Turn off flashlight",
    ActionType.PLAY_YOUTUBE:      lambda p: "Open YouTube",
    ActionType.PLAY_MUSIC:        lambda p: "Play music",
    ActionType.OPEN_CAMERA:       lambda p: "Open camera",
    ActionType.OPEN_GALLERY:      lambda p: "Open gallery",
    ActionType.ENABLE_BLUETOOTH:  lambda p: "Enable Bluetooth",
    ActionType.DISABLE_BLUETOOTH: lambda p: "Disable Bluetooth",
    ActionType.ENABLE_DND:        lambda p: "Enable Do Not Disturb",
    ActionType.DISABLE_DND:       lambda p: "Disable Do Not Disturb",
    ActionType.CALL_CONTACT:      lambda p: f"Call {p.get('contact', 'unknown')}",
    ActionType.SEND_SMS:          lambda p: f"Send message to {p.get('contact', 'unknown')} saying {p.get('message', '')}",
    ActionType.OPEN_APP:          lambda p: f"Open {p.get('app', 'app')}",
    ActionType.SET_BRIGHTNESS:    lambda p: f"Set brightness to {p.get('value', '50')}%",
    ActionType.SET_VOLUME:        lambda p: f"Set volume to {p.get('value', '50')}%",
    ActionType.SET_ALARM:         lambda p: f"Set alarm for {p.get('time', '7:00 AM')}",
    ActionType.SET_TIMER:         lambda p: f"Set timer for {p.get('minutes', '10')} minutes",
    ActionType.COPY_TEXT:         lambda p: f"Copy {p.get('text', '')}",
    ActionType.SHOW_NOTIFICATION: lambda p: f"Show notification {p.get('message', '')}",
    ActionType.NOVA_CORE_REQUEST: lambda p: p.get('request', 'Nova Core request'),
    ActionType.WAIT_SECONDS:      lambda p: "Wait",
}


class ActionLog:
    def __init__(self, action_type: str, is_success: bool,
                 spoken_response: str, error_message: Optional[str] = None,
                 duration_ms: int = 0):
        self.action_type = action_type
        self.is_success = is_success
        self.spoken_response = spoken_response
        self.error_message = error_message
        self.duration_ms = duration_ms


class ActionExecutor:
    def __init__(self, command_engine=None):
        self.command_engine = command_engine

    def execute(self, actions: List[AutomationAction]) -> List[ActionLog]:
        logs = []
        for action in actions:
            start = time.time()
            cmd_text = self._build_text(action)
            logger.info(f"Executing: {action.type} → \"{cmd_text}\"")

            try:
                if self.command_engine and action.type != ActionType.WAIT_SECONDS:
                    result = self.command_engine.execute_text(cmd_text)
                    log = ActionLog(
                        action_type=action.type.value,
                        is_success=result.is_success,
                        spoken_response=result.spoken_response,
                        error_message=result.error_message,
                        duration_ms=int((time.time() - start) * 1000)
                    )
                else:
                    # WAIT_SECONDS or no engine → simulate
                    wait = action.delay_seconds_after or int(action.params.get("seconds", 0))
                    if wait > 0:
                        time.sleep(min(wait, 5))  # cap test waits at 5s
                    log = ActionLog(action.type.value, True, f"Executed {action.type.value}",
                                   duration_ms=int((time.time() - start) * 1000))
            except Exception as e:
                logger.error(f"Action {action.type} failed: {e}", exc_info=True)
                log = ActionLog(action.type.value, False, "",
                                error_message=str(e),
                                duration_ms=int((time.time() - start) * 1000))
            logs.append(log)

            if action.delay_seconds_after > 0 and action.type != ActionType.WAIT_SECONDS:
                time.sleep(min(action.delay_seconds_after, 2))  # cap in tests

        return logs

    def _build_text(self, action: AutomationAction) -> str:
        builder = ACTION_TEXT_MAP.get(action.type)
        return builder(action.params) if builder else action.type.value

File: mobile/automation/ActionExecutor.kt
package com.nova.mobile.automation

import android.util.Log
import com.nova.mobile.command.CommandEngine
import kotlinx.coroutines.delay

/**
 * ActionExecutor — Executes a list of AutomationActions sequentially.
 * Reuses the existing CommandEngine to translate actions into plugin calls.
 * Supports per-action delays and error isolation (one failure doesn't abort all).
 */
class ActionExecutor(private val commandEngine: CommandEngine) {
    companion object {
        private const val TAG = "ActionExecutor"
        private const val ACTION_TIMEOUT_MS = 8000L
    }

    data class ActionLog(
        val actionType: String,
        val isSuccess: Boolean,
        val spokenResponse: String,
        val errorMessage: String? = null,
        val durationMs: Long = 0L
    )

    suspend fun execute(actions: List<AutomationAction>): List<ActionLog> {
        val log = mutableListOf<ActionLog>()
        for (action in actions) {
            val start = System.currentTimeMillis()
            val text = buildCommandText(action)
            Log.i(TAG, "Executing action: ${action.type} → \"$text\"")

            val actionLog = try {
                val result = commandEngine.executeText(text)
                ActionLog(
                    actionType = action.type.name,
                    isSuccess = result.isSuccess,
                    spokenResponse = result.spokenResponse,
                    errorMessage = result.errorMessage,
                    durationMs = System.currentTimeMillis() - start
                )
            } catch (e: Exception) {
                Log.e(TAG, "Action ${action.type} failed: ${e.message}")
                ActionLog(
                    actionType = action.type.name,
                    isSuccess = false,
                    spokenResponse = "",
                    errorMessage = e.message,
                    durationMs = System.currentTimeMillis() - start
                )
            }
            log.add(actionLog)

            if (action.delaySecondsAfter > 0) {
                Log.i(TAG, "Waiting ${action.delaySecondsAfter}s before next action…")
                delay(action.delaySecondsAfter * 1000L)
            }
        }
        return log
    }

    /** Translates an AutomationAction into a natural-language string the CommandEngine understands. */
    private fun buildCommandText(action: AutomationAction): String = when (action.type) {
        ActionType.CALL_CONTACT     -> "Call ${action.params["contact"] ?: "unknown"}"
        ActionType.SEND_SMS         -> "Send message to ${action.params["contact"] ?: "unknown"} saying ${action.params["message"] ?: ""}"
        ActionType.OPEN_APP         -> "Open ${action.params["app"] ?: "app"}"
        ActionType.PLAY_YOUTUBE     -> "Open YouTube"
        ActionType.PLAY_MUSIC       -> "Play music"
        ActionType.FLASHLIGHT_ON    -> "Turn on flashlight"
        ActionType.FLASHLIGHT_OFF   -> "Turn off flashlight"
        ActionType.SET_BRIGHTNESS   -> "Set brightness to ${action.params["value"] ?: "50"}%"
        ActionType.SET_VOLUME       -> "Set volume to ${action.params["value"] ?: "50"}%"
        ActionType.ENABLE_BLUETOOTH -> "Enable Bluetooth"
        ActionType.DISABLE_BLUETOOTH-> "Disable Bluetooth"
        ActionType.ENABLE_DND       -> "Enable Do Not Disturb"
        ActionType.DISABLE_DND      -> "Disable Do Not Disturb"
        ActionType.SET_ALARM        -> "Set alarm for ${action.params["time"] ?: "7:00 AM"}"
        ActionType.SET_TIMER        -> "Set timer for ${action.params["minutes"] ?: "10"} minutes"
        ActionType.OPEN_CAMERA      -> "Open camera"
        ActionType.OPEN_GALLERY     -> "Open gallery"
        ActionType.COPY_TEXT        -> "Copy ${action.params["text"] ?: ""}"
        ActionType.SHOW_NOTIFICATION-> "Show notification ${action.params["message"] ?: ""}"
        ActionType.NOVA_CORE_REQUEST-> action.params["request"] ?: "Nova Core request"
        ActionType.WAIT_SECONDS     -> "Wait"
    }
}
